Source/WebCore/page/SecurityOriginData.h

```diff
@@ -25,36 +25,52 @@
 
 #pragma once
 
+#include "ProcessQualified.h"
+#include <wtf/ArgumentCoder.h>
 #include <wtf/Hasher.h>
+#include <wtf/Markable.h>
 #include <wtf/URL.h>
 
 namespace WebCore {
 
 class Frame;
 class SecurityOrigin;
 
+enum class ShouldSkipOpaqueSecurityOriginCheck : bool { No, Yes };
+
 struct SecurityOriginData {
+    enum OpaqueOriginIdentifierType { };
+    using OpaqueOriginIdentifier = ProcessQualified<ObjectIdentifier<OpaqueOriginIdentifierType>>;
+    using MarkableOpaqueOriginIdentifier = Markable<OpaqueOriginIdentifier, OpaqueOriginIdentifier::MarkableTraits>;
+
     SecurityOriginData() = default;
     SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port)
         : protocol(protocol)
         , host(host)
         , port(port)
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
+        if ((protocol.isEmpty() || protocol.isNull()) && (host.isEmpty() || host.isNull()))
```

Comment:
a null SecurityOriginData is different from an opaque SecurityOriginData. Therefore, setting *this to an opaque origin  when protocol && host are null seems wrong to me.

Reply:
Also, seems we should check that `!port` to maintain pre-existing behavior.